src/arch/cuda/hybridAPI/buddy_allocator.h

```diff
@@ -0,0 +1,247 @@
+    int bucket_count;
+
+    // Map of allocated blocks
+    std::unordered_map<uint8_t*, AllocBlock> alloc_map;
+
+    /* --------------------- */
+    /* | Utility functions | */
+    /* --------------------- */
+
+    void print_status() {
+      printf("(buckets)\n");
+      size_t free = 0;
+      for (int i = 0; i < bucket_count; i++) {
+        printf("bucket[%d]: ", i);
+        for (const auto& block : buckets[i]) {
+          free += block.size;
+          printf("{%p, %lu} ", block.ptr, block.size);
+        }
+        printf("\n");
+      }
+
+      printf("(alloc_map)\n");
+      size_t allocated = 0;
+      size_t used = 0;
+      for (const auto& elem : alloc_map) {
+        const auto& block = elem.second;
+        allocated += block.size;
+        used += block.requested;
+        printf("ptr: %p, size: %lu, req: %lu\n", elem.first, block.size, block.requested);
+      }
+
+      printf("(fragmentation) free: %lu, allocated: %lu, used: %lu\n", free, allocated, used);
+    }
+
+    size_t get_free_size() {
+      size_t free = 0;
+      for (int i = 0; i < bucket_count; i++) {
+        for (const auto& block : buckets[i]) {
+          free += block.size;
+        }
+      }
+
+      return free;
+    }
+
+    int get_bucket(size_t size) {
+      return (int)std::ceil(std::log2((double)size)) - 2;
+    }
+
+    int get_block_index(uint8_t* ptr, size_t size) {
+      return (size_t)(ptr - base_ptr) / size;
+    }
+
+    /* ------------------------ */
+    /* | Allocation functions | */
+    /* ------------------------ */
+
+    allocator(size_t size) : min_alloc(4), base_ptr(NULL) {
+      if (size <= 0) {
+        fprintf(stderr, "Allocator size has to be larger than 0 bytes\n");
+        abort();
+      }
+
+      // Request GPU memory (closest power of 2)
+      int limit_log2 = std::ceil(std::log2((double)size));
+      limit = (size_t)std::pow(2, limit_log2);
+      cudaMalloc(&base_ptr, limit);
+      DEBUG_PRINT("Initialized base_ptr %p with %lu bytes\n", (void*)base_ptr, limit);
+
+      // Initialize buckets and set up last bucket (for size min_alloc)
+      bucket_count = limit_log2 - 1;
+      buckets = new std::list<FreeBlock>[bucket_count];
+      buckets[bucket_count-1].emplace_back(base_ptr, limit);
+    }
+
+    ~allocator() {
+      // Free GPU memory
+      cudaFree(base_ptr);
+      delete[] buckets;
+    }
+
+    void* malloc(size_t request) {
+      // Cannot satisfy request larger than limit
+      if (request > limit) return nullptr;
+
+      // Has to be larger than minimum allocation size (4 bytes)
+      // Size is rounded up to the nearest power of 2
+      size_t alloc_size = (request > min_alloc) ? request : min_alloc;
+      int bucket = get_bucket(alloc_size);
+      int original_bucket = bucket;
+
+      // Find an empty bucket
+      while (buckets[bucket].empty() && bucket < bucket_count) {
+        bucket++;
+      }
+
+      // All buckets were empty!
+      if (bucket == bucket_count) {
```

Comment:
Should the comment read "// All buckets were full!"?

Reply:
In the case where no free blocks are available, I'm wondering if we should allocate more device memory using `cudaMalloc` and keep all the allocator objects in a list? I'm guessing if multiple consecutive allocations are requested and all of them wait to be freed (because of transfers in progress), then more memory cannot be requested since the buckets will be full?

Reply:
> Should the comment read "// All buckets were full!"?

You're right, will change this

> In the case where no free blocks are available, I'm wondering if we should allocate more device memory using `cudaMalloc` and keep all the allocator objects in a list? I'm guessing if multiple consecutive allocations are requested and all of them wait to be freed (because of transfers in progress), then more memory cannot be requested since the buckets will be full?

Currently more memory is not allocated when there are no free blocks because the underlying memory is mmap'ed (for use of CUDA IPC) and cannot be dynamically resized. I'm planning to update this so that we can allocate a larger backing store once it runs out and mmap it again (albeit costly, because it has to be mmap'ed on all processes).